### DISCUSSION TITLE
make typegoose like trails/tynamo

### DISCUSSION THREAD

**Discussion by nhhockeyplayer** [typegoose/typegoose#652] [Ideas]
I wanted to suggest maybe take a peek at a framework I use to be a committer on back in 2006/2007 founded by Chris Nelson founder of java.net he recruited me to help on that.

The framework was "trails" at the time. Its now called tynamo handled by 2-4 developers. It was the sweetest archetype on the planet at the time and only for special interest. Some of the best data structure modelings were cast within that framework. It was an end to end CRUD archetype. Hardwired to hibernate and tapestry sadly but the core feature was reflection with fundamental PROPERTY DESCRIPTORS that offered a turnkey fullstack project generator.

What did the developer get after creating a new project?

Out of the box a complete running fullstack app based on java, wired to hibernate and presented by tapestry UI.

Hint Hint... typegoose, wired to mongoose and presented by IONIC with a default set of UI Pages for free out of the box as depicted in https://github.com/tynamo/tapestry-model
and why not go the mile and do this one too for **_property based_** security and **_instance based_** security https://github.com/tynamo/tapestry-security

But under the hood was something really neat... the PropertyDescriptor decorator and more of the same variations of such. Those custom decorators were applied static out of the box to a base set of CRUD user interface modules that come with every model/entity LIST, EDIT, VIEW, ... well I might be off but here is the main core lib

https://github.com/tynamo/tapestry-model

and with that you got what it did under the hood free UI (configurable skins included) for any model you dreamed of having in your fullstack end to end. These free UI pages came generated out of the archetype but by the several @PropertyDescriptor decorators that go heavy on reflection by analyzing each class (your actual model(s)), its properties and methods and automagic UI came free right out of the box.

What attracted me to typegoose was https://github.com/szokodiakos brilliant use of reflection and decorators

typegoose framework is already there and just needs that tiny push in the right direction. I created a ticket years back for typegoose as an enhancement to make it a platform just like tynamo/trails.

It would be nice to see this happen and awesome if it could fall directly onto IONIC

The use of reflect-metadata in typegoose is impressive but it does not go as far as what tynamo/trails did by bundling a set of CRUD/List/Edit/Query/View/Find static default pages like tynanmo/trails does and settling on a simple UI framework.

The PropertyDescriptor is the key construct in this all

If the typegoose team can seriously take a peek at this typegoose is almost there it just needs a push.

It would push typegoose as the elite framework on the planet as the constructs in tynamo/trails are the best on the planet if you take a peek at those they really are hard to resist.

They have several other libs optional to integrate like https://github.com/tynamo/tapestry-security and even a REST one also... the security lib https://github.com/tynamo/tapestry-security is pants wetting its so good... instance based security and property based security for any model.

Im not kidding these constructs would be the takeover of the entire planet of this space if integrated to typegoose and it wouldnt take much to accomplish either.

HibernateDescriptorDecorator was one part of that code core and central mechanics also

it would be nice to see this happen.

typegoose would never be caught

I highly encourage looking into this and the team seriously considering.

The property descriptor decorator and its variations/derivatives would be the way to start

**Comment by hasezoey** [comment]
for future readers, this discussion is basically the same message as in https://github.com/typegoose/typegoose/issues/638#issuecomment-1003481794 , i already had answered there, but will do it again here

----

if i understand the project `https://github.com/tynamo/tapestry-model` correctly (i am not a java developer), then it is basically what typegraphql / typeorm / typegoose does, but also adds a generator for the models (like [swagger](https://www.npmjs.com/package/swagger) does for a rest api) and also some extra default provided models (unlike the default classes typegoose already provides)

but this is out-of-scope for the (main, this repository here) typegoose project - though i would not be against if someone creates something like that and would be added to https://github.com/typegoose/

**Reply by nhhockeyplayer** [reply]
the kool thing about trails/tynamo is mechanically speaking... the app is actually (once generated) a mirror of itself... an amazing symmetric (and I would go as far as 'sacred') outcome. it has to be looked at to really see what the end result really is after reflect-meta data gets thru with the code

hard to wrap your head around what im describing but if anyone chooses to look and see it may hit conceptually and you will see what Im talking about.

whereby each property on a class (like ID) would have its own custom @PropertyDescriptor decorator that pans out to its widget markup editor on the page for the UI... and this would carry on for all primitive native types but as types of properties get complex like array we end up with @ListPropertyDescriptor all of which could be subclassed and customized by developers based on their complexity.

And then the ORM property descriptors come into play @OneToOne @OneToMany @ManyToOne @ManyToMany and their respective markup UI template projected in on generation

you would have to generate a maven project from their archetypes to see it in action its impressive

**Comment by smolinari** [comment]
So, I've got most of the POCs done for a beginning system, something like Tapestry, and my ideas go a lot further than theirs have I think. I still have a loooonnngggg way to go though, to get something done that can offer even a Tapestry-like experience. It's a dream of mine, to build a programming and business application platform. One, where devs only need to deal with complicated business logic and UI components and/ or integrations, but can also offer pre-built apps (as a service), so business people can just plug-n-play those apps, or extend on them with their own devs. 

Btw, I'm using Quasar Framework for the UI, which, if you ask me, is a step up from Ionic. :) 

Scott
